### ISSUE TITLE
Input: Cloudwatch - Specify dimensions to include (regardless of value)

### ISSUE THREAD

**Issue by johnrengelman** [influxdata/telegraf#1115]
Currently the cloudwatch input requires you to specify the KV Dimension value for a cloudwatch metric, otherwise it explicitly ignores Dimensions which results in Cloudwatch returning aggregate data.
The only way to get the plugin to process data for each individual entity would be to explicitly list each Dimension value which is not realistic for a dynamic environment.

Let's assume that we have 2 ELBs (elb1, elb2), that are both deployed in  2 AZs (us-east-1a, us-east-1b).
In this scenario, Cloud watch actually has 7 metrics for each MetricName (i.e. RequestCount)
- Dimensions = [], aggregate across all 2 ELBs & 2 AZs
- Dimensions = [az=us-east-1a], aggregate across 2 ELBs in us-east-1a
- Dimensions = [az=us-east-1b], aggregate across 2 ELBs in us-east-1b
- Dimensions = [name=elb1, az=us-east-1a], data for ELB1 in us-east-1a only
- Dimensions = [name=elb2, az=us-east-1a], data for ELB2 only in us-east-1a only
- Dimensions = [name=elb1, az=us-east-1b], data for ELB1 only in us-east-1b only
- Dimensions = [name=elb2, az=us-east-1b], data for ELB2 only in us-east-1b only

The desired output here is to export metrics by name & az, but _not_ the global, az, and name aggregations.
This requires configuring the `Dimension` value on the input explicitly for the names of name and az.
In a dynamic environment, this might not be known.
Instead, we should be able to specify which dimensions to select, list the available metrics, selecting only the ones that contain all the specified dimensions (regardless of value) and then retrieve metric data using those dimension filters.

**Comment by sparrc** [comment]
@ljosa @joshhardy thoughts on this?

**Comment by joshhardy** [comment]
@sparrc - as @johnrengelman mentioned, the current plugin supports either very explicitly specified metrics, or *all* the things. 
That met requirements for @ljosa and for me, but I agree that more flexibility would be nice. 

@johnrengelman - you want to specify something like this?
```toml
  [[inputs.cloudwatch.metrics]]
    names = ['Latency', 'RequestCount']

    [[inputs.cloudwatch.metrics.dimensions]]
      name = 'LoadBalancerName'
```
Which would result in the plugin exporting a metric for each available load balancer name (as though you had listed each explicitly in the config)?

**Comment by johnrengelman** [comment]
@joshhardy yeah, that's right.

Looking at the code it changes the flow a little bit because you would need to do a `ListMetrics` regardless of the configuration now, so you can get the available dimension values. (there's no API for acquiring the unique dimension sets).

**Comment by joshhardy** [comment]
@johnrengelman - we would either have to give up that optimization to skip the `ListMetrics`, or do some fairly wonky config parsing to decide if it was needed or not.
Since there is a long-lived cache for the ListMetrics value, it is not a huge hit to make that call regardless, and that opens the door to more flexible config options.
The current implementation only hits `ListMetrics` once per hour.

It seems like fair tradeoff to me.
Thoughts?

**Comment by johnrengelman** [comment]
Yeah...I think making the CacheTTL configurable and defaulting to 1 hr would make that a fine tradeoff.

**Comment by danielnelson** [comment]
Closed in #1208
